eth.ts

```diff
@@ -2,30 +2,69 @@ import Web3 from "web3";
 import {compiledContracts} from "./compiled-contracts";
 import { Contract as Web3Contract } from "web3-eth-contract";
 import BN from "bn.js";
+import { Contracts } from "./typings/contracts";
 const HDWalletProvider = require("truffle-hdwallet-provider");
 
 export const ETHEREUM_URL = process.env.ETHEREUM_URL || "http://localhost:7545";
 const ETHEREUM_MNEMONIC = process.env.ETHEREUM_MNEMONIC || "vanish junk genuine web seminar cook absurd royal ability series taste method identify elevator liquid";
 
+export const defaultWeb3Provider = () => new Web3(new HDWalletProvider(
+    ETHEREUM_MNEMONIC,
+    ETHEREUM_URL,
+    0,
+    100,
+    false
+    ));
 
-const refreshWeb3 = () => {
-    web3 = new Web3(new HDWalletProvider(
-        ETHEREUM_MNEMONIC,
-        ETHEREUM_URL,
-        0,
-        100,
-        false
-        )
-    );
-};
+export class Web3Driver{
+    private web3 : Web3;
+    private nonceVersion = 0;
+    constructor(private web3Provider : () => Web3 = defaultWeb3Provider){
+        this.web3 = this.web3Provider();
+    }
 
-export var web3;
-refreshWeb3();
+    get eth(){
+        return this.web3.eth;
+    }
+    get currentProvider(){
+        return this.web3.currentProvider;
+    }
+
+    async deploy<N extends keyof Contracts>(contractName: N, args: any[], options?: any) {
+        try {
+            const abi = compiledContracts[contractName].abi;
+            const accounts = await this.web3.eth.getAccounts();
+            let web3Contract = await new this.web3.eth.Contract(abi).deploy({
+                data: compiledContracts[contractName].bytecode,
+                arguments: args || []
+            }).send({
+                from: accounts[0],
+                ...(options || {})
+            });
+            let lastNonce = this.nonceVersion;
+            const web3ContractProvider = () => {
```

Comment:
these two are really complicating things here. 
what if you had a contract dictionary object in this instance mapping addresses to the underlying contract object? then, instead of passing `web3ContractProvider` to `Contract` constructor you passed only the contract's address - the key to the dictionary that holds all and only "fresh" contract objects.

then, in the getter `Contract.web3Contract(): Web3Contract` you would use the address to check if `web3` (the driver) has a fresh contract object for that address. if it does not, the contract object initializes it and registers it so it will be there next time we look (replacing line 47) and only then returns the object to the caller.

and lastly, whenever we replace the web3 object we simply replace the contract dictionary with an empty object - essentially forcing all contract objects to be flushed. what do you think?

Reply:
I had a similar model at some point (with worse model). IMO it will make the code a bit more readable but also a bit more complex. definitely more Object Oriented. no objection, I will adjust in a separate commit.

Reply:
done, see b7b901c